can/interfaces/slcan.py

```diff
@@ -161,7 +162,6 @@ def _read(self, timeout: Optional[float]) -> Optional[str]:
             while not (
                 ord(self._OK) in self._buffer or ord(self._ERROR) in self._buffer
             ):
-                self.serialPortOrig.timeout = time_left
```

Comment:
Why did you remove this?

Reply:
Setting the serial port timeout in the while loop slow down the RX performance.

Reply:
Is there a way to ensure, that the function respects the `timeout` parameter?

Reply:
> Is there a way to ensure, that the function respects the `timeout` parameter?

- This is the test code for testing the `timeout` parameter

```py
import time

from can.interfaces.slcan import slcanBus
from can.message import Message

bus = slcanBus(channel="COM11", bitrate=20000, sleep_after_open=0)

tx_msg = Message(
    arbitration_id=0x3C,
    is_extended_id=False,
    is_remote_frame=False,
    dlc=8,
    data=bytes(
        [
            0x01,
            0x02,
            0x10,
            0x01,
            0xFF,
            0xFF,
            0xFF,
            0xFF,
        ]
    ),
)

rx_msg = Message(
    arbitration_id=0x3D,
    is_extended_id=False,
    is_remote_frame=True,
    dlc=8,
)

count = 5
timeout = 1

start_time = time.time()
for _ in range(count):
    bus.send(tx_msg)
    bus.send(rx_msg)
    msg = bus.recv(timeout)
    assert msg is None
    print(msg)
end_time = time.time()
spend_time = end_time - start_time
assert abs((timeout * count) - spend_time) < 0.1
print(f"{spend_time:.3f}")

bus.shutdown()
```

- This is the result
![image](https://user-images.githubusercontent.com/41532794/213179419-a038485f-18b7-4e55-b98c-fc6e20c0c69c.png)

Reply:
With this proposed change I think there's still a scenario where the timeout is not enforced.

Reply:
Would this work?

```python3
if timeout is not None:
    self.serialPortOrig.timeout = timeout
self._buffer += self.serialPortOrig.read_until(self._OK)
```

Reply:
It would work for expected messages but not errors I believe?

But I wrote a unittest that accurately catches the original error. Going to open a separate PR.

Reply:
> Would this work?
> 
> ```python
> if timeout is not None:
>     self.serialPortOrig.timeout = timeout
> self._buffer += self.serialPortOrig.read_until(self._OK)
> ```

hi, @zariiii9003 

This reminds me that the pyserial has an RX buffer inside. 

I think we can remove `self._buffer` if we use `read_until`.

This code works and it's simpler and more effective than the previous one

```py
def _read(self, timeout: Optional[float]) -> Optional[str]:

    with error_check("Could not read from serial device"):
        if timeout is not None:
            self.serialPortOrig.timeout = timeout
        buffer = self.serialPortOrig.read_until(self._OK)
        if buffer:
            return buffer.strip().decode()
```

Reply:
As @Tbruno25 said, we would still need to check whether there is a `self._ERROR` in the returned buffer. There even might be a `self._ERROR` and a valid message after the error.

Reply:
While reviewing the `pyserial` [docs](https://pyserial.readthedocs.io/en/latest/pyserial_api.html#serial.Serial.__init__) I noticed `timeout` has a non-blocking mode 

> timeout = 0: non-blocking mode, return immediately in any case, returning zero or more, up to the requested number of bytes

@mikisama It seems like setting the serial port to non-blocking before the while loop and adjusting accordingly may achieve the results you're after.

Reply:
@Tbruno25 

I try your suggestion and I found it causes higher CPU usage.

The I7-7700HQ has about 13-18% CPU usage, while the previous blocking mode had about 3% CPU usage.

```py
# pyserial non-blocking mode
self.serialPortOrig.timeout = 0
# pyserial blocking mode
# self.serialPortOrig.timeout = timeout
# first read what is already in receive buffer
while self.serialPortOrig.in_waiting:
    self._buffer += self.serialPortOrig.read()
# if we still don't have a complete message, do a blocking read
start = time.time()
time_left = timeout
while not (
    ord(self._OK) in self._buffer or ord(self._ERROR) in self._buffer
):
    byte = self.serialPortOrig.read()
    if byte:
        self._buffer += byte
    ...
```

Reply:
> As @Tbruno25 said, we would still need to check whether there is a `self._ERROR` in the returned buffer. There even might be a `self._ERROR` and a valid message after the error.

Thanks for the explanation, I now understand what @Tbruno25 means

Reply:
@Tbruno25 

Following your thoughts, I found maybe we can try `read_all`.

- The `read_all` method of `pyserial` is no-blocking too, but it also causes higher CPU usage. 
- The advantage of using the `read_all` method is that we don't have to deal with the `timeout` parameter of `serialPortOrig`

```py
with error_check("Could not read from serial device"):
    # with `read_all`, we don't have to deal with the `timeout` parameter of `serialPortOrig`
    # self.serialPortOrig.timeout = timeout
    if self.serialPortOrig.in_waiting:
        self._buffer += self.serialPortOrig.read_all()
    # if we still don't have a complete message, do a blocking read
    start = time.time()
    time_left = timeout
    while not (
        ord(self._OK) in self._buffer or ord(self._ERROR) in self._buffer
    ):
        byte = self.serialPortOrig.read_all()
        if byte:
            self._buffer += byte
        ...
```

Reply:
I'm not sure I agree. `read_all` appears to be omitted from the pyserial docs, but from looking at the source code it calls the underlying `read` function.

```
    def read_all(self):
        """\
        Read all bytes currently available in the buffer of the OS.
        """
        return self.read(self.in_waiting)
```

This would still be subject to violating the timeout.

Reply:
@Tbruno25 

If there is no data in the buffer, `in_waiting` is 0, and `read(0)` will return immediately.

```py
def read_all(self):
    return self.read(self.in_waiting)
```

You can try this code.

```py
import serial
import time

with serial.Serial("COM11") as ser:
    start = time.time()
    for _ in range(100):
        data = ser.read_all()
        # make sure read_all get nothing
        assert data == bytes()
    end = time.time()
    delta = end - start
    # check whether read_all is no-blocking 
    assert delta < 0.01
    print(f"{delta:.3f}")
```
And the result.
![image](https://user-images.githubusercontent.com/41532794/214758398-1e7fafe5-2b47-4acb-b581-84d181e7eb88.png)

Reply:
But what about when there _is_ data in the buffer?

Please reference the test case in #1492 for an example.

Reply:
`read_all` sounds good to me.

```python3
    def _read(self, timeout: Optional[float]) -> Optional[str]:
        _timeout = serial.Timeout(timeout)

        with error_check("Could not read from serial device"):
            while not _timeout.expired():
                new_data = self.serialPortOrig.read_all()
                if new_data is not None:
                    self._buffer.extend(new_data)
                else:
                    time.sleep(0.001)
                    continue

                for terminator in (self._ERROR, self._OK):
                    if terminator in self._buffer:
                        i = self._buffer.index(terminator) + 1
                        string = self._buffer[:i].decode()
                        del self._buffer[:i]
                        return string

            return None
```

The serial port must be instantiated with `timeout=0`.

Reply:
> But what about when there _is_ data in the buffer? What if there were an infinite amount of 'junk'?

@Tbruno25 

```py
def read(self, __size: int = ...) -> bytes: ...
def read_all(self):
    return self.read(self.in_waiting)
```
1. If there is data in the buffer, the `read` method only copies it out. Only when the data bytes in the buffer are smaller than the parameter `__size`, the `read` method will blocking until the data bytes in the buffer is `__size`, then copy it out.
2. Since the `in_waiting` parameter shows the current number of bytes in the buffer, blocking will not occur. (The time to copy data out is negligibly small)
3. According to my test, the `MAX` receive buffer size in win10 is `16384`, and ubuntu 20.04 is `4095`. I'm not sure whether it can be changed. But I think there will not be `infinite` junk.

```py
import serial
import time

with serial.Serial("COM11") as ser:
    # send a start command to make
    # the STM32 dev board start sending data.
    ser.write(b"start")

    while True:
        time.sleep(0.5)
        print(f"in_waiting {ser.in_waiting}")
```
![image](https://user-images.githubusercontent.com/41532794/215014478-b40e9937-74da-43ff-a8f2-6f50f6291436.png)

![image](https://user-images.githubusercontent.com/41532794/215013165-fef1ec0d-0d80-4a53-afb4-6d4f030243b5.png)

Reply:
@zariiii9003 

Looks good. Just a few minor changes requested.

According to my test, the `read_all` method returns `b''` if there is no data to be read out.

So the `if new_data is not None:` should be changed to `if new_data:`, otherwise it will never do `time.sleep`, which will cause high CPU usage.

```py
    def _read(self, timeout: Optional[float]) -> Optional[str]:
        _timeout = serial.Timeout(timeout)

        with error_check("Could not read from serial device"):
            while not _timeout.expired():
                new_data = self.serialPortOrig.read_all()
                if new_data:
                    self._buffer.extend(new_data)
                else:
                    time.sleep(0.001)
                    continue

                for terminator in (self._ERROR, self._OK):
                    if terminator in self._buffer:
                        i = self._buffer.index(terminator) + 1
                        string = self._buffer[:i].decode()
                        del self._buffer[:i]
                        return string

            return None
```

 >The serial port must be instantiated with timeout=0.

By the way, I did not initialize the `timeout` parameter of `serialPortOrig` at all and it defaults to `None`. It seems still works for me. 

I'm curious. If you don't set the `timeout` to 0, what will happen to you?

Reply:
> I'm curious. If you don't set the timeout to 0, what will happen to you?

Nothing, i don't have an interface to test this 😄

Reply:
> According to my test, the MAX receive buffer size in win10 is 16384, and ubuntu 20.04 is 4095. I'm not sure whether it can be changed. But I think there will not be infinite junk.

Interesting! And I agree -- I wasn't considering the physical limitations of the hardware/kernel. Good catch :slightly_smiling_face: